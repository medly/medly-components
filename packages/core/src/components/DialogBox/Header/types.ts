import { HTMLProps, WithThemeProp } from '@medly-components/utils';
export interface DialogBoxHeaderProps extends HTMLProps<HTMLDivElement>, WithThemeProp {
    id: string;
}
