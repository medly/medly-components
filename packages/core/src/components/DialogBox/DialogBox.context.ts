import { createContext } from 'react';
import { DialogBoxContextType } from './types';

// default values are for testing purposes
export const defaultValues: DialogBoxContextType = {
    id: 'default-id'
};

export const DialogBoxContext = createContext<DialogBoxContextType>(defaultValues);
