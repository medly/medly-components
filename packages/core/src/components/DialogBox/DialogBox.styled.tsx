import { rgba } from 'polished';
import styled from 'styled-components';
import { DialogBoxBackgroundProps } from './types';

export const DialogBoxBackgroundStyled = styled.div<DialogBoxBackgroundProps>`
    display: flex;
    justify-content: center;
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: ${({ theme }) => theme.modal.overlayColor};
    z-index: 1000;
    align-items: center;
    animation: ${({ open }) => `${open ? 'bgFadeIn' : 'bgFadeOut'} 0.4s cubic-bezier(0, 0, 0.33, 1) `};
    overflow: hidden;

    @keyframes bgFadeIn {
        0% {
            background: ${({ theme }) => rgba(theme.modal.overlayColor, 0)};
        }
        100% {
            background: ${({ theme }) => rgba(theme.modal.overlayColor, 0.65)};
        }
    }

    @keyframes bgFadeOut {
        0% {
            background: ${({ theme }) => rgba(theme.modal.overlayColor, 0.65)};
        }
        100% {
            background: ${({ theme }) => rgba(theme.modal.overlayColor, 0)};
        }
    }
`;
