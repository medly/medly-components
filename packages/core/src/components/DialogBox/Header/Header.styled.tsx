import { styled } from '@medly-components/utils';
import { DialogBoxHeaderProps } from './types';


export const Header = styled('div')<DialogBoxHeaderProps>`
    background-color: ${({ theme }) => theme.modal.backgroundColor};
    color: ${({ theme }) => theme.modal.headerColor};
    font-weight: 500;
    margin-bottom: 2.4rem;
`;
