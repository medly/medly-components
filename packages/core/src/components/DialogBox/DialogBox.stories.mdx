import { DialogBox } from './DialogBox';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './DialogBox.stories';

<Meta title="Core" component={DialogBox} parameters={{ jest: ['DialogBox.test.tsx'] }} />

# DialogBox

`DialogBox` are windows that appear in front of app content to give critical information, request a decision, or involve multiple tasks.  It consists of a Header, Content, and Action.

You can play with the component in the canvas tab.

```typescript
<DialogBox open={openState} onCloseDialogBox={changeDialogBoxState(false)}>
    <DialogBox.Header>Add User</DialogBox.Header>
    <DialogBox.Content>
        <Input id="name-input" type="text" fullWidth label="Name" placeholder="Enter your Name" />
    </DialogBox.Content>
    <DialogBox.Actions>
        <Button variant="outlined">Cancel</Button>
    </DialogBox.Actions>
</DialogBox>
```

Shadow appears on the Header and Actions section if you scroll through the content.

To close the dialog box, you can either:

1. Click the cross icon.
2. Press `Esc` key.
3. If the `shouldCloseOnOutsideClick` prop is provided, then the dialog box will close on clicking outside the dialog box.

<Preview withToolbar>
    <Story name="DialogBox">{stories.Basic()}</Story>
</Preview>

## Props

This component offers the following props mentioned in the table below:

<Props of={DialogBox} />

## DialogBox Action Props

The DialogBox action prop is given in the table below:

<Props of={stories.DialogBoxActionProps} />

## DialogBoxTheme

This component offers the following themes mentioned in the table below:

<Props of={stories.ThemeInterface} />
