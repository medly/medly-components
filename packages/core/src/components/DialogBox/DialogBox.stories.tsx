import { defaultTheme, DialogBoxTheme } from '@medly-components/theme';
import { useCallback, useState } from 'react';
import Button from '../Button';
import { DialogBoxActionUserProps } from './Actions/types';
import { DialogBox } from './DialogBox';
import type { FC } from 'react';

export const ThemeInterface: FC<DialogBoxTheme> = () => null;
ThemeInterface.defaultProps = {
    ...defaultTheme.modal
};

export const DialogBoxActionProps: FC<DialogBoxActionUserProps> = () => null;
DialogBoxActionProps.defaultProps = {
    alignItems: 'right'
};

export const Basic = () => {
    const [modalState, setModalState] = useState(false);

    const changeModalState = useCallback(() => setModalState(val => !val), []);

    return (
        <>
            <Button onClick={changeModalState}>Click to Open</Button>
            <DialogBox open={modalState} onCloseModal={changeModalState} shouldCloseOnOutsideClick={true} showCloseIcon={true}>
                <DialogBox.Header>Are you sure?</DialogBox.Header>
                <DialogBox.Content>
                    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna
                    aliqua.
                </DialogBox.Content>
                <DialogBox.Actions>
                    <Button edges="rounded">Delete</Button>
                    <Button variant="outlined" edges="rounded" onClick={() => changeModalState()}>
                        Cancel
                    </Button>
                </DialogBox.Actions>
            </DialogBox>
        </>
    );
};
