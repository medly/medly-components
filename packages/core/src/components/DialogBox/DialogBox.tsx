import { useCombinedRefs, useKeyPress, WithStyle } from '@medly-components/utils';
import React, { FC, useCallback, useEffect, useState } from 'react';
import Actions from './Actions';
import Content from './Content';
import { DialogBoxContext } from './DialogBox.context';
import { DialogBoxBackgroundStyled } from './DialogBox.styled';
import Header from './Header';
import Popup from './Popup';
import { DialogBoxProps, DialogBoxStaticProps } from './types';

const Component: FC<DialogBoxProps> = React.memo(
    React.forwardRef((props, ref) => {
        const { id, open, onCloseModal, children, minWidth, shouldCloseOnOutsideClick, minHeight, ...restProps } = props,
            isEscPressed = useKeyPress('Escape'),
            dialogBoxRef = useCombinedRefs<HTMLDivElement>(ref, React.useRef(null)),
            [shouldRender, setShouldRender] = useState(open);

        const handleBackgroundClick = useCallback(() => {
            shouldCloseOnOutsideClick && onCloseModal && onCloseModal();
        }, [shouldCloseOnOutsideClick, onCloseModal]);

        useEffect(() => {
            if (open) {
                setShouldRender(true);
                document.body.style.overflow = 'hidden';
            }
            return () => {
                document.body.style.overflow = 'unset';
            };
        }, [open]);

        const handleAnimationEnd = useCallback(() => {
            if (!open) {
                setShouldRender(false);
                document.body.style.overflow = 'unset';
            }
        }, [open]);

        useEffect(() => {
            open && isEscPressed && onCloseModal && onCloseModal();
        }, [open, isEscPressed]);

        return shouldRender ? (
            <DialogBoxBackgroundStyled {...{ ...restProps, id, open }} onClick={handleBackgroundClick}>
                <Popup ref={dialogBoxRef} id={`${id}-popup`} onAnimationEnd={handleAnimationEnd} {...{ minWidth, minHeight, open }}>
                    <DialogBoxContext.Provider value={{ id: id! }}>{children}</DialogBoxContext.Provider>
                </Popup>
            </DialogBoxBackgroundStyled>
        ) : null;
    })
);

Component.defaultProps = {
    id: 'medly-dialog-box',
    open: false,
    shouldCloseOnOutsideClick: false
};
Component.displayName = 'DialogBox';
export const DialogBox: FC<DialogBoxProps> & WithStyle & DialogBoxStaticProps = Object.assign(Component, {
    Style: DialogBoxBackgroundStyled,
    Header,
    Popup,
    Content,
    Actions
});
