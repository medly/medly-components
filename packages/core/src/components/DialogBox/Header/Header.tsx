import { isValidStringOrNumber, WithStyle } from '@medly-components/utils';
import type { FCC } from 'react';
import { Children, memo, useContext } from 'react';
import Text from '../../Text';
import { DialogBoxContext } from '../DialogBox.context';
import * as Styled from './Header.styled';

const Component: FCC = memo(({ children }) => {
    const { id } = useContext(DialogBoxContext);

    return (
        <Styled.Header {...{ id: `${id}-header` }}>
            {Children.map(children, c => {
                return isValidStringOrNumber(c) ? <Text textVariant="h4">{c}</Text> : c;
            })}
        </Styled.Header>
    );
});
Component.displayName = 'Header';
export const Header: FCC & WithStyle = Object.assign(Component, { Style: Styled.Header });
