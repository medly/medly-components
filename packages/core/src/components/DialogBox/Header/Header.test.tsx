import { render } from '@test-utils';
import { defaultValues, DialogBoxContext } from '../DialogBox.context';
import { Header } from './Header';

const DialogBoxHeaderRenderer = () => {
    return render(
        <DialogBoxContext.Provider value={{ ...defaultValues }}>
            <Header>test</Header>
        </DialogBoxContext.Provider>
    );
};

describe('DialogBox component header', () => {
    it('should render properly at small screen size', () => {
        const { container } = DialogBoxHeaderRenderer();
        expect(container).toMatchSnapshot();
    });
});
