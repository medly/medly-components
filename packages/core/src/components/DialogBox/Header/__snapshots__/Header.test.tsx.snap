// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogBox component header should render properly at small screen size 1`] = `
.c1 {
  margin: 0;
  color: inherit;
  font-size: 1.8rem;
  font-weight: 700;
  letter-spacing: -0.02rem;
  line-height: 2.4rem;
  text-align: initial;
}

.c0 {
  background-color: #ffffff;
  color: #13181D;
  font-weight: 500;
  margin-bottom: 2.4rem;
}

<div>
  <div
    class="c0"
    id="default-id-header"
  >
    <h4
      class="c1"
      textalign="initial"
      textcolor="inherit"
      textvariant="h4"
    >
      test
    </h4>
  </div>
</div>
`;
